designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,288 @@
+The base writer class uses function call operator overloading
+to hand the following kinds of output data:
+
++ a vector of strings (names)
++ a vector of doubles (data)
++ a string (message)
++ nothing ()
+
+The `stan::services` layer provides a set of calling functions which
+wrap the available inference algorithms.
+Stan interfaces instantiate the `writer` object and pass it into the `services` layer method
+which in turn passes this along to the inference algorithm.
+The calling functions for the HMC sampler and ADVI provide two arguments slots for
+`stan::callbacks::writer` used for CSV output files:
+the CSV file contining the sampler outputs and a 
+[diagnostic_file](https://mc-stan.org/docs/2_29/cmdstan-guide/mcmc-config.html#sampler-diag-file)
+which contains the per-iteration latent Hamiltonian dynamics of all model parameters,
+as described here:  https://discourse.mc-stan.org/t/sampler-hmc-diagnostics-file/15386;
+the optimization algorithms only provide a single output CSV file slot.
+
+The CmdStan interface instantiates the `stan::callbacks::writer` object for the Stan CSV file
+and writes a series of header comment lines which record the full set of CmdStan arguments as well
+as configuration information about both the Stan version and the model object.
+The remainder of the Stan CSV file contents are generated by the `services` layer method.
+
+
+## Functional Specification
+
+The results from one run of the Stan inference engine will be factored into
+a series of information and type-specific outputs.
+An end-user will be able to specify the outputs of interest.
+The interfaces will have control over the output streams.
+For a file-based iterface, such as CmdStan, filesystem directories will be used to organize
+the resulting set of output files, e.g., instead of command-line argument `file=output.csv`,
+the interface will have argument `directory=output`.
+
+
+### Kinds of outputs, suggested names
+
+Files which are updated with each iteration (i.e., streaming data), which can be monitored by downstream processes:
+
+- `sample` - A table containing the per-draw estimates of all model variables on the constrained scale; 1 column per variable (element), one row per iteration.
+
+- `log_prob` - A table containing the value of `lp__`, the joint log probability density, one row per iteration.
+- `algorithm_state` - A table containing the per-iteration state of the inference algorithm, e.g. `accept_stat__`, `log_p__`, one row per iteration.
+
+Note that if we adjoin all the columns in tables `log_prob`, `algorithm_state`, and `_sample`, the resulting data table would contain the equivalent information
+in the data table rows of Stan CSV file.
+
+Files which contain global information, produced once per inference algorithm run:
+
+- `stepsize` - A single element table containing the stepsize.
+- `metric` - A table which contains the inverse mass matrix arrived at by adaptation.
```

Comment:
yes, they definitely could be in a single JSON file and this would be the right thing to do because it would allow for re-use - that's the argument I was looking for.